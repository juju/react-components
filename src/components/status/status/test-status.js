/* Copyright (C) 2017 Canonical Ltd. */

import React from 'react';
import enzyme from 'enzyme';

import Status from './status';

describe('Status', () => {
  let model, valueStore;

  const renderComponent = (options = {}) =>
    enzyme.shallow(
      <Status
        generateApplicationOnClick={
          options.generateApplicationOnClick === undefined
            ? sinon.stub()
            : options.generateApplicationOnClick
        }
        generateApplicationURL={
          options.generateApplicationURL === undefined
            ? sinon.stub()
            : options.generateApplicationURL
        }
        generateCharmURL={
          options.generateCharmURL === undefined ? sinon.stub() : options.generateCharmURL
        }
        generateMachineOnClick={
          options.generateMachineOnClick === undefined
            ? sinon.stub()
            : options.generateMachineOnClick
        }
        generateMachineURL={
          options.generateMachineURL === undefined ? sinon.stub() : options.generateMachineURL
        }
        generateUnitOnClick={
          options.generateUnitOnClick === undefined
            ? sinon.stub()
            : options.generateUnitOnClick
        }
        generateUnitURL={
          options.generateUnitURL === undefined ? sinon.stub() : options.generateUnitURL
        }
        getIconPath={options.getIconPath || sinon.stub()}
        model={options.model || model}
        navigateToApplication={
          options.navigateToApplication === undefined
            ? sinon.stub()
            : options.navigateToApplication
        }
        navigateToCharm={
          options.navigateToCharm === undefined ? sinon.stub() : options.navigateToCharm
        }
        navigateToMachine={
          options.navigateToMachine === undefined ? sinon.stub() : options.navigateToMachine
        }
        valueStore={options.valueStore || valueStore}
      />
    );

  beforeEach(() => {
    model = {
      cloud: 'aws',
      environmentName: 'my-model',
      modelUUID: 'myuuid',
      region: 'neutral zone',
      sla: 'advanced',
      version: '2.42.47'
    };
    valueStore = {
      applications: {
        app: {
          status: {
            current: ''
          }
        }
      },
      machines: {
        machine: {
          agentStatus: {
            current: ''
          }
        }
      },
      relations: {relation: {}},
      remoteApplications: {app: {}},
      units: {
        unit: {
          agentStatus: {
            current: ''
          },
          workloadStatus: {
            current: ''
          }
        }
      }
    };
  });

  it('renders when not connected to a model', () => {
    model = {};
    const wrapper = renderComponent();
    assert.equal(
      wrapper.find('.status-view__content').text(),
      'Cannot show the status: the GUI is not connected to a model.'
    );
  });

  it('renders without entities', () => {
    valueStore.applications = {};
    valueStore.machines = {};
    valueStore.relations = {};
    valueStore.remoteApplications = {};
    valueStore.units = {};
    const wrapper = renderComponent();
    expect(wrapper).toMatchSnapshot();
  });

  it('renders with entites', () => {
    const wrapper = renderComponent();
    expect(wrapper).toMatchSnapshot();
  });

  it('renders without link functions', () => {
    const wrapper = renderComponent({
      generateApplicationOnClick: null,
      generateApplicationURL: null,
      generateCharmURL: null,
      generateMachineOnClick: null,
      generateMachineURL: null,
      generateUnitOnClick: null,
      generateUnitURL: null,
      navigateToApplication: null,
      navigateToCharm: null,
      navigateToMachine: null
    });
    expect(wrapper).toMatchSnapshot();
  });

  it('can set the highest status', () => {
    valueStore.units = {
      unit: {
        agentStatus: {
          current: 'error'
        },
        workloadStatus: {
          current: ''
        }
      }
    };
    const wrapper = renderComponent();
    assert.equal(wrapper.find('StatusModel').prop('highestStatus'), 'error');
  });
});
