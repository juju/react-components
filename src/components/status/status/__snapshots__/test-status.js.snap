// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Status renders with entites 1`] = `
<Panel
  focus={true}
  instanceName="status-view"
  visible={true}
>
  <div
    className="status-view__content"
  >
    <StatusModel
      changeFilter={[Function]}
      counts={
        Object {
          "applications": 1,
          "machines": 1,
          "relations": 1,
          "remoteApplications": 1,
          "units": 1,
        }
      }
      highestStatus="ok"
      model={
        Object {
          "cloud": "aws",
          "environmentName": "my-model",
          "modelUUID": "myuuid",
          "region": "neutral zone",
          "sla": "advanced",
          "version": "2.42.47",
        }
      }
      statusFilter={null}
    />
    <StatusRemoteApplicationList
      remoteApplications={
        Object {
          "app": Object {},
        }
      }
      statusFilter={null}
    />
    <StatusApplicationList
      applications={
        Object {
          "app": Object {
            "status": Object {
              "current": "",
            },
          },
        }
      }
      generateApplicationOnClick={[Function]}
      generateApplicationURL={[Function]}
      generateCharmURL={[Function]}
      getIconPath={[Function]}
      onCharmClick={[Function]}
      statusFilter={null}
      units={
        Object {
          "unit": Object {
            "agentStatus": Object {
              "current": "",
            },
            "workloadStatus": Object {
              "current": "",
            },
          },
        }
      }
    />
    <StatusUnitList
      applications={
        Object {
          "app": Object {
            "status": Object {
              "current": "",
            },
          },
        }
      }
      generateMachineURL={[Function]}
      generateUnitOnClick={[Function]}
      generateUnitURL={[Function]}
      getIconPath={[Function]}
      onMachineClick={[Function]}
      statusFilter={null}
      units={
        Object {
          "unit": Object {
            "agentStatus": Object {
              "current": "",
            },
            "workloadStatus": Object {
              "current": "",
            },
          },
        }
      }
    />
    <StatusMachineList
      generateMachineOnClick={[Function]}
      generateMachineURL={[Function]}
      machines={
        Object {
          "machine": Object {
            "agentStatus": Object {
              "current": "",
            },
          },
        }
      }
      statusFilter={null}
    />
    <StatusRelationList
      applications={
        Object {
          "app": Object {
            "status": Object {
              "current": "",
            },
          },
        }
      }
      generateApplicationURL={[Function]}
      getIconPath={[Function]}
      onApplicationClick={[Function]}
      relations={
        Object {
          "relation": Object {},
        }
      }
      statusFilter={null}
    />
  </div>
</Panel>
`;

exports[`Status renders without entities 1`] = `
<Panel
  focus={true}
  instanceName="status-view"
  visible={true}
>
  <div
    className="status-view__content"
  >
    <StatusModel
      changeFilter={[Function]}
      counts={
        Object {
          "applications": 0,
          "machines": 0,
          "relations": 0,
          "remoteApplications": 0,
          "units": 0,
        }
      }
      highestStatus="ok"
      model={
        Object {
          "cloud": "aws",
          "environmentName": "my-model",
          "modelUUID": "myuuid",
          "region": "neutral zone",
          "sla": "advanced",
          "version": "2.42.47",
        }
      }
      statusFilter={null}
    />
  </div>
</Panel>
`;

exports[`Status renders without link functions 1`] = `
<Panel
  focus={true}
  instanceName="status-view"
  visible={true}
>
  <div
    className="status-view__content"
  >
    <StatusModel
      changeFilter={[Function]}
      counts={
        Object {
          "applications": 1,
          "machines": 1,
          "relations": 1,
          "remoteApplications": 1,
          "units": 1,
        }
      }
      highestStatus="ok"
      model={
        Object {
          "cloud": "aws",
          "environmentName": "my-model",
          "modelUUID": "myuuid",
          "region": "neutral zone",
          "sla": "advanced",
          "version": "2.42.47",
        }
      }
      statusFilter={null}
    />
    <StatusRemoteApplicationList
      remoteApplications={
        Object {
          "app": Object {},
        }
      }
      statusFilter={null}
    />
    <StatusApplicationList
      applications={
        Object {
          "app": Object {
            "status": Object {
              "current": "",
            },
          },
        }
      }
      generateApplicationOnClick={null}
      generateApplicationURL={null}
      generateCharmURL={null}
      getIconPath={[Function]}
      onCharmClick={null}
      statusFilter={null}
      units={
        Object {
          "unit": Object {
            "agentStatus": Object {
              "current": "",
            },
            "workloadStatus": Object {
              "current": "",
            },
          },
        }
      }
    />
    <StatusUnitList
      applications={
        Object {
          "app": Object {
            "status": Object {
              "current": "",
            },
          },
        }
      }
      generateMachineURL={null}
      generateUnitOnClick={null}
      generateUnitURL={null}
      getIconPath={[Function]}
      onMachineClick={null}
      statusFilter={null}
      units={
        Object {
          "unit": Object {
            "agentStatus": Object {
              "current": "",
            },
            "workloadStatus": Object {
              "current": "",
            },
          },
        }
      }
    />
    <StatusMachineList
      generateMachineOnClick={null}
      generateMachineURL={null}
      machines={
        Object {
          "machine": Object {
            "agentStatus": Object {
              "current": "",
            },
          },
        }
      }
      statusFilter={null}
    />
    <StatusRelationList
      applications={
        Object {
          "app": Object {
            "status": Object {
              "current": "",
            },
          },
        }
      }
      generateApplicationURL={null}
      getIconPath={[Function]}
      onApplicationClick={null}
      relations={
        Object {
          "relation": Object {},
        }
      }
      statusFilter={null}
    />
  </div>
</Panel>
`;
